src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,643 @@
+    /// Fixed column used in scalar decomposition for variable-base scalar mul
+    pub mul_decompose: Column<Fixed>,
+
+    /// Point doubling
+    pub q_double: Selector,
+    /// Incomplete addition
+    pub q_add_incomplete: Selector,
+    /// Complete addition
+    pub q_add: Selector,
+    /// Variable-base scalar multiplication (hi half)
+    pub q_mul_hi: Selector,
+    /// Variable-base scalar multiplication (lo half)
+    pub q_mul_lo: Selector,
+    /// Variable-base scalar multiplication (final scalar)
+    pub q_mul_decompose: Selector,
+    /// Fixed-base full-width scalar multiplication
+    pub q_mul_fixed: Selector,
+    /// Fixed-base signed short scalar multiplication
+    pub q_mul_fixed_short: Selector,
+    /// Witness point
+    pub q_point: Selector,
+    /// Witness full-width scalar for fixed-base scalar mul
+    pub q_scalar_fixed: Selector,
+    /// Witness signed short scalar for full-width fixed-base scalar mul
+    pub q_scalar_fixed_short: Selector,
+    /// Permutation
+    pub perm: Permutation,
+}
+
+/// A chip implementing EccInstructions
+#[derive(Clone, Debug)]
+pub struct EccChip<C: CurveAffine> {
+    id: u64,
+    config: EccConfig,
+    loaded: EccLoaded<C>,
+}
+
+impl<C: CurveAffine> PartialEq for EccChip<C> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl<C: CurveAffine> Eq for EccChip<C> {}
+
+impl<C: CurveAffine> EccLoaded<C> {
+    fn get(&self, point: OrchardFixedBases<C>) -> OrchardFixedBase<C> {
+        match point {
+            OrchardFixedBases::CommitIvkR(_) => self.commit_ivk_r.clone(),
+            OrchardFixedBases::NoteCommitR(_) => self.note_commit_r.clone(),
+            OrchardFixedBases::NullifierK(_) => self.nullifier_k.clone(),
+            OrchardFixedBases::ValueCommitR(_) => self.value_commit_r.clone(),
+        }
+    }
+
+    fn get_short(&self, _point: OrchardFixedBasesShort<C>) -> OrchardFixedBaseShort<C> {
+        self.value_commit_v.clone()
+    }
+}
+
+impl<C: CurveAffine> Chip<C::Base> for EccChip<C> {
+    type Config = EccConfig;
+    type Loaded = EccLoaded<C>;
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &self.loaded
+    }
+}
+
+impl<C: CurveAffine> EccChip<C> {
+    pub fn construct(
+        config: <Self as Chip<C::Base>>::Config,
+        loaded: <Self as Chip<C::Base>>::Loaded,
+    ) -> Self {
+        Self {
+            id: rand::random::<u64>(),
+            config,
+            loaded,
+        }
+    }
+
+    #[allow(non_snake_case)]
+    #[allow(clippy::many_single_char_names)]
+    #[allow(clippy::too_many_arguments)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<C::Base>,
+        bits: Column<Advice>,
+        P: (Column<Advice>, Column<Advice>),
+        lambda: (Column<Advice>, Column<Advice>),
+        extras: [Column<Advice>; 5],
+    ) -> <Self as Chip<C::Base>>::Config {
+        let q_double = meta.selector();
```

Comment:
We don't use doubling in the Orchard circuit; how much cost are we paying for this selector?

Reply:
The doubling API, module, and selector have now been flagged off with `#[cfg(test)]`.

Reply:
The doubling API has now been removed. I've opened a ticket to reintroduce it at some later time (see #90)